core/headerchain.go

```diff
@@ -312,6 +331,16 @@ func (hc *HeaderChain) GetTd(hash common.Hash) *big.Int {
 	return td
 }
 
+// WriteTd stores a block's total difficulty into the database, also caching it
+// along the way.
+func (hc *HeaderChain) WriteTd(hash common.Hash, td *big.Int) error {
+	if err := WriteTd(hc.chainDb, hash, td); err != nil {
+		return err
+	}
+	hc.tdCache.Add(hash, new(big.Int).Set(td))
```

Comment:
I'm afraid that some code now or in the future might reuse TD calculations, which will then modify the cached value. This way it's future proof that the thing in the cache is unique and you don't need to worry.

The semantics is the same as we had till now with writing it out to the database and reading back a copy, just lighter.